File: backend/logic/importer.py
import pandas as pd
import numpy as np
import io
import re
from typing import Dict, List, Any, Optional, Tuple
from datetime import datetime
from dateutil import parser
import hashlib

def parse_file_to_kpis(file_content: bytes, filename: str, months_count: int = 12) -> Dict[str, Any]:
    """
    Parses an uploaded file (Excel or CSV) into all its raw sheets.
    Returns a dictionary mapping sheet names to lists of lists.
    """
    is_excel = filename.endswith('.xlsx') or filename.endswith('.xls')
    sheets_data = {}
    
    if is_excel:
        excel_file = pd.ExcelFile(io.BytesIO(file_content))
        for sheet_name in excel_file.sheet_names:
            df = pd.read_excel(excel_file, sheet_name=sheet_name, header=None)
            df_clean = df.replace([np.inf, -np.inf], np.nan)
            raw_rows = df_clean.where(pd.notnull(df_clean), None).values.tolist()
            sheets_data[sheet_name] = raw_rows
    else:
        # CSV Handling (Single sheet by definition)
        try:
            df = pd.read_csv(io.BytesIO(file_content), header=None, encoding='utf-8')
        except UnicodeDecodeError:
            df = pd.read_csv(io.BytesIO(file_content), header=None, encoding='latin1')
            
        df_clean = df.replace([np.inf, -np.inf], np.nan)
        raw_rows = df_clean.where(pd.notnull(df_clean), None).values.tolist()
        sheets_data["Default"] = raw_rows

    return sanitize_for_json({
        "sheets": sheets_data,
        "kpis": {}, # Empty for backward compatibility during transition
        "raw_rows": sheets_data.get(next(iter(sheets_data))) if sheets_data else []
    })

def sanitize_for_json(obj: Any) -> Any:
    """Recursively replaces non-JSON compliant floats (NaN, Inf) with None."""
    if isinstance(obj, dict):
        return {k: sanitize_for_json(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [sanitize_for_json(x) for x in obj]
    elif isinstance(obj, float):
        if np.isnan(obj) or np.isinf(obj):
            return None
        return obj
    return obj

def _find_col(header: List[str], markers: List[str], default: int = -1) -> int:
    for marker in markers:
        if marker in header:
            return header.index(marker)
    return default

def _extract_number(val: Any) -> Optional[float]:
    if pd.isna(val) or val == "":
        return None
    
    # Common zero markers in financial data
    financial_zeros = ["-", "--", "( - )", "n.a.", "n/a", "tbd", "actuals", "target"]
    if isinstance(val, str) and val.strip().lower() in financial_zeros:
        return 0.0

    if isinstance(val, (int, float)):
        return float(val)
    
    s = str(val).strip()
    
    # Handle negative brackets
    if s.startswith('(') and s.endswith(')'):
        inner = s[1:-1].strip()
        if inner in ["-", "--"]: return 0.0
        s = '-' + inner
    
    # Handle currency symbols and common separators
    s = s.replace('$', '').replace('£', '').replace('€', '')
    if ',' in s and '.' in s:
        s = s.replace(',', '')
    elif ',' in s and s.count(',') == 1:
        parts = s.split(',')
        if len(parts[1]) != 3: s = s.replace(',', '.')
        else: s = s.replace(',', '')
    else:
        s = s.replace(',', '')

    s = re.sub(r'[^0-9.\-eE%]', '', s)
    if not s or s == '-' or s == '.':
        return None
        
    try:
        if s.endswith('%'):
            return float(s[:-1]) / 100.0
        return float(s)
    except ValueError:
        return None

def parse_excel_date(val: Any) -> Optional[datetime]:
    if isinstance(val, (int, float)) and val > 1000:
        try:
            return pd.to_datetime(val, unit='D', origin='1899-12-30').to_pydatetime()
        except:
             return None
    return None

def parse_date_to_absolute_month(date_str: Any, fallback_year: int = 2024) -> Optional[int]:
    if date_str is None or date_str == "":
        return None
    
    m_idx = -1
    y_idx = -1
    
    # 1. Handle Excel Serial Dates
    excel_dt = parse_excel_date(date_str)
    if excel_dt:
        m_idx = excel_dt.month - 1
        y_idx = excel_dt.year
    
    # 2. Handle Numbers (Month Indices)
    elif isinstance(date_str, (int, float)):
        val = int(date_str)
        if 1 <= val <= 12: m_idx = val - 1
        elif 0 <= val <= 11: m_idx = val
        if m_idx != -1: y_idx = fallback_year
    
    # 3. Handle String Dates
    else:
        try:
            dt = parser.parse(str(date_str), fuzzy=True)
            m_idx = dt.month - 1
            y_idx = dt.year
        except:
            s = str(date_str).lower()
            months = ["jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"]
            for i, month in enumerate(months):
                if month in s:
                    m_idx = i
                    break
            
            # Improved year detection to handle "feb2025" (no boundary between letters and numbers)
            year_match = re.search(r'(?<!\d)(20\d{2})(?=\b)', s)
            if not year_match:
                year_match = re.search(r'(20\d{2})', s)
            
            if year_match:
                y_val = int(year_match.group(1))
                y_idx = y_val if y_val > 100 else 2000 + y_val
            else:
                # Try 2-digit year fallback if separated or at end
                short_year_match = re.search(r'(\d{2})$', s)
                if short_year_match:
                    y_idx = 2000 + int(short_year_match.group(1))
                else:
                    y_idx = fallback_year

    if m_idx != -1 and y_idx != -1:
        return y_idx * 12 + m_idx
    return None

def promote_by_hierarchy(rows: List[List[Any]], months_count: int, mappings: Dict[str, str], 
                         start_month: int = 0, start_year: int = 2024) -> Dict[str, Any]:
    """
    Builds a structural KPI tree from flat data.
    Supports both Wide format (months in columns) and Long format (Time, Scenario, Value columns).
    """
    if not rows or len(rows) < 2:
        return {}

    header = [str(c).strip() if c else "" for c in rows[0]]
    data_rows = rows[1:]
    
    # 1. Identify Column Roles
    roles = {} # role -> index
    hierarchy_levels = [] # list of (level, index) e.g., [(1, 0), (2, 1)]
    
    if mappings:
        for idx_str, role in mappings.items():
            try:
                idx = int(idx_str)
                if role.startswith("L"):
                    try:
                        level = int(role[1:])
                        hierarchy_levels.append((level, idx))
                    except: pass
                else:
                    roles[role] = idx
            except: pass
    else:
        # Heuristic discovery
        for i, col in enumerate(header):
            c_upper = col.upper()
            if c_upper == "L1" or "LEVEL 1" in c_upper: hierarchy_levels.append((1, i))
            elif c_upper == "L2" or "LEVEL 2" in c_upper: hierarchy_levels.append((2, i))
            elif c_upper == "L3" or "LEVEL 3" in c_upper: hierarchy_levels.append((3, i))
            elif c_upper == "L4" or "LEVEL 4" in c_upper: hierarchy_levels.append((4, i))
            elif any(x in c_upper for x in ["SCENARIO", "VERSION", "TYPE", "DATA TYPE", "ACTUAL/BUDGET"]): 
                roles["Scenario"] = i

    if not hierarchy_levels:
        # Default to first few columns as hierarchy if nothing else, but SKIP scenario/time/value columns
        possible_indices = [i for i in range(len(header)) if i not in roles.values()]
        if len(possible_indices) >= 2:
            hierarchy_levels = [(1, possible_indices[0]), (2, possible_indices[1])]
        elif len(possible_indices) == 1:
            hierarchy_levels = [(1, possible_indices[0])]
        else:
            hierarchy_levels = [(1, 0), (2, 1)] # Fallback

    hierarchy_levels.sort() # Ensure L1 comes before L2

    # 2. Determine Format (Wide vs Long) and Auto-Detect Date Range bounds
    has_months_in_header = any(parse_date_to_absolute_month(c, start_year) is not None for c in header)
    is_long_format = "Value" in roles and ("Time" in roles or has_months_in_header)
    
    # Collect all absolute months present in the dataset
    time_abs_values = []
    
    if is_long_format:
        time_idx = roles.get("Time")
        if time_idx is not None:
            for row in data_rows:
                if time_idx < len(row):
                    abs_m = parse_date_to_absolute_month(row[time_idx], start_year)
                    if abs_m is not None:
                        time_abs_values.append(abs_m)
    else:
        for i, col_name in enumerate(header):
            if i in [idx for _, idx in hierarchy_levels] or i == roles.get("Scenario"):
                continue
            abs_m = parse_date_to_absolute_month(col_name, start_year)
            if abs_m is not None:
                time_abs_values.append(abs_m)
                
    detected_date_range = None
    if time_abs_values:
        min_abs_month = min(time_abs_values)
        max_abs_month = max(time_abs_values)
        
        start_year = min_abs_month // 12
        start_month = min_abs_month % 12
        end_year = max_abs_month // 12
        end_month = max_abs_month % 12
        
        span = max_abs_month - min_abs_month + 1
        
        # Don't strictly shrink if the user requested a larger grid (e.g. for padding/forecasting)
        months_count = max(span, months_count)
        
        # Recalculate end_month and end_year to match the final padded months_count
        final_abs_month = min_abs_month + months_count - 1
        end_year = final_abs_month // 12
        end_month = final_abs_month % 12
        
        detected_date_range = {
            "startMonth": start_month,
            "startYear": start_year,
            "endMonth": end_month,
            "endYear": end_year,
            "monthsCount": months_count
        }
    
    # 3. Process Rows into a Hierarchy Map
    # Structure: { kpi_key: { id: str, label: str, values: { scenario: [values] } } }
    kpi_map = {}
    
    def get_kpi_key(levels_data: List[str]) -> str:
        return " > ".join(levels_data)

    for row in data_rows:
        # Extract hierarchy
        current_levels = []
        for _, idx in hierarchy_levels:
            if idx < len(row):
                val = str(row[idx]).strip() if row[idx] is not None else ""
                current_levels.append(val)
        
        if not any(current_levels):
            continue # Skip blank rows
            
        # Clean current_levels (trailing blanks)
        while current_levels and not current_levels[-1]:
            current_levels.pop()
            
        if not current_levels:
            continue
            
        kpi_key = get_kpi_key(current_levels)
        if kpi_key not in kpi_map:
            # Generate a stable ID based on the hierarchy path
            kpi_id = f"kpi-{hashlib.md5(kpi_key.encode()).hexdigest()[:16]}"
            kpi_map[kpi_key] = {
                "id": kpi_id,
                "label": current_levels[-1],
                "levels": current_levels,
                "scenarios": {} # scenario_name -> [0] * months_count
            }
        
        scenario_name = "Actual"
        if "Scenario" in roles and roles["Scenario"] < len(row):
            scenario_name = str(row[roles["Scenario"]]).strip() or "Actual"

        if scenario_name not in kpi_map[kpi_key]["scenarios"]:
            kpi_map[kpi_key]["scenarios"][scenario_name] = [0.0] * months_count

        # Extract values
        if is_long_format:
            # Long format logic
            val_idx = roles.get("Value")
            time_idx = roles.get("Time")
            
            val = _extract_number(row[val_idx]) if val_idx is not None and val_idx < len(row) else 0.0
            
            m_idx = 0
            if time_idx is not None and time_idx < len(row):
                abs_m = parse_date_to_absolute_month(row[time_idx], start_year)
                if abs_m is not None:
                    m_idx = abs_m - (start_year * 12 + start_month)
                
            if 0 <= m_idx < months_count:
                kpi_map[kpi_key]["scenarios"][scenario_name][m_idx] += val
        else:
            # Wide format logic (months are columns)
            for i, col_name in enumerate(header):
                # Skip hierarchy/scenario columns
                if i in [idx for _, idx in hierarchy_levels] or i == roles.get("Scenario"):
                    continue
                
                abs_m = parse_date_to_absolute_month(col_name, start_year)
                if abs_m is not None:
                    m_idx = abs_m - (start_year * 12 + start_month)
                    if 0 <= m_idx < months_count:
                        v = _extract_number(row[i]) if i < len(row) else 0.0
                        kpi_map[kpi_key]["scenarios"][scenario_name][m_idx] += v

    # 4. Build Relationship Map & Registry
    # We do this twice: once to establish structure, then again per scenario to populate data.
    node_registry = {}
    sorted_keys = sorted(kpi_map.keys(), key=lambda x: len(x.split(" > ")))
    
    # First pass: structure only
    for key in sorted_keys:
        item = kpi_map[key]
        levels = item["levels"]
        for i in range(1, len(levels) + 1):
            sub_path = levels[:i]
            sub_key = " > ".join(sub_path)
            if sub_key not in node_registry:
                p_path = levels[:i-1]
                p_key = " > ".join(p_path) if p_path else None
                
                # Use the same stable ID logic
                node_id = f"kpi-{hashlib.md5(sub_key.encode()).hexdigest()[:16]}"
                
                node_registry[sub_key] = {
                    "id": node_id,
                    "label": sub_path[-1],
                    "unit": "$",
                    "formula": "SUM" if i < len(levels) else "NONE",
                    "children": [],
                    "parentId": node_registry[p_key]["id"] if p_key in node_registry else None,
                    "data": [], # Will be populated per scenario
                    "isExpanded": True,
                    "simulationValue": 0,
                    "simulationType": "PERCENT",
                }
    
    # Link children (Existing logic handles the CSV hierarchy, we just appended scenario nodes)
    for node in node_registry.values():
        p_id = node["parentId"]
        if p_id:
            for p_node in node_registry.values():
                if p_node["id"] == p_id:
                    if node["id"] not in p_node["children"]:
                        p_node["children"].append(node["id"])
                    break

    # 5. Populate Data for each scenario
    all_scenarios = set()
    for item in kpi_map.values():
        all_scenarios.update(item["scenarios"].keys())
    
    if not all_scenarios:
        all_scenarios = {"Actual"}

    scenarios_output = {}
    for s_name in all_scenarios:
        # Create a full set of KPIs for this scenario
        s_kpis = {}
        # 5.1 Initialize leaf data for this scenario
        for key, node in node_registry.items():
            s_node = node.copy()
            s_node["children"] = node["children"].copy()
            
            # Find data in kpi_map if it's a leaf we actually have data for
            vals = [0.0] * months_count
            if key in kpi_map and s_name in kpi_map[key]["scenarios"]:
                vals = kpi_map[key]["scenarios"][s_name]
            
            # Calculate actual month and year corresponding to `j`
            s_node["data"] = [
                {
                    "month": (start_year * 12 + start_month + j) % 12,
                    "year": (start_year * 12 + start_month + j) // 12,
                    "actual": v
                } 
                for j, v in enumerate(vals)
            ]
            s_kpis[s_node["id"]] = s_node

        # 5.2 Aggregation: Sum up parent data from children for this scenario
        depth_sorted_keys = sorted(node_registry.keys(), key=lambda x: len(x.split(" > ")), reverse=True)
        for key in depth_sorted_keys:
            node_id = node_registry[key]["id"]
            s_node = s_kpis[node_id]
            if s_node["children"]:
                # We want to aggregate ONLY from children that are NOT scenario nodes.
                regular_children = []
                for cid in s_node["children"]:
                    child_node_key = next((k for k, v in node_registry.items() if v["id"] == cid), None)
                    # If this child is a scenario node (its key ends with a scenario name), skip it
                    if child_node_key and " > " in child_node_key and child_node_key.split(" > ")[-1] in all_scenarios:
                        continue
                    regular_children.append(cid)
                    
                if regular_children:
                    totals = [0.0] * months_count
                    for cid in regular_children:
                        if cid in s_kpis:
                            c_data = s_kpis[cid]["data"]
                            for j in range(min(len(c_data), months_count)):
                                totals[j] += c_data[j]["actual"]
                    
                    s_node["data"] = [
                        {
                            "month": (start_year * 12 + start_month + j) % 12,
                            "year": (start_year * 12 + start_month + j) // 12,
                            "actual": v
                        }
                        for j, v in enumerate(totals)
                    ]
        
        # Also handle scenario child nodes if created
        for s_key, s_item in kpi_map.items():
            for s_name_inner, s_values in s_item["scenarios"].items():
                scen_node_key = f"{s_key} > {s_name_inner}"
                if scen_node_key in node_registry:
                    sn_id = node_registry[scen_node_key]["id"]
                    s_kpis[sn_id] = node_registry[scen_node_key].copy()
                    s_kpis[sn_id]["isScenarioNode"] = True
                    s_kpis[sn_id]["data"] = [
                        {"month": (start_year*12+start_month+j)%12, "year": (start_year*12+start_month+j)//12, "actual": v}
                        for j, v in enumerate(s_values)
                    ]
        
        scenarios_output[s_name] = s_kpis

    # 6. Return Result
    stats = {
        "total_rows_processed": len(data_rows),
        "kpis_created": len(kpi_map),
        "scenarios_detected": list(all_scenarios)
    }

    res = {}
    if len(all_scenarios) > 1 or (len(all_scenarios) == 1 and next(iter(all_scenarios)) != "Actual"):
        default_scenario = "Actual" if "Actual" in scenarios_output else \
                          ("Base" if "Base" in scenarios_output else next(iter(scenarios_output)))
        res = {
            "type": "multi-scenario",
            "scenarios": scenarios_output,
            "kpis": scenarios_output[default_scenario]
        }
    else:
        res = scenarios_output.get("Actual", scenarios_output.get(next(iter(scenarios_output))))
    
    if isinstance(res, dict):
        res["stats"] = stats
        if detected_date_range:
            res["dateRange"] = detected_date_range
            # Also ensure startMonth/startYear in the root result match detected values for UI
            res["startMonth"] = detected_date_range["startMonth"]
            res["startYear"] = detected_date_range["startYear"]
    return res

File: application/src/logic/Importer.ts
import * as XLSX from 'xlsx';

import type { KPIData, TimeSeriesValue, DateRange } from '../types';

export type KPIValue = TimeSeriesValue;
export type KPINode = KPIData;

/**
 * Generates a stable ID for a KPI based on its hierarchy path.
 */
function getStableId(path: string): string {
    let hash = 0;
    for (let i = 0; i < path.length; i++) {
        const char = path.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash; // Convert to 32bit integer
    }
    return `kpi-${Math.abs(hash).toString(16).padStart(8, '0')}`;
}

export function parseDateToAbsoluteMonth(dateStr: any, fallbackYear: number = 2024): number | null {
    if (dateStr === null || dateStr === undefined || dateStr === "") return null;

    let mIdx = -1;
    let yIdx = -1;

    // 1. Handle Numbers (Excel serial dates or simple indices)
    if (typeof dateStr === 'number') {
        if (dateStr > 1000) {
            // Excel Serial Date fallback logic
            const dt = new Date(Math.round((dateStr - 25569) * 86400 * 1000));
            mIdx = dt.getUTCMonth();
            yIdx = dt.getUTCFullYear();
        } else {
            if (dateStr >= 1 && dateStr <= 12) mIdx = dateStr - 1;
            else if (dateStr >= 0 && dateStr <= 11) mIdx = dateStr;
            if (mIdx !== -1) yIdx = fallbackYear;
        }
    } else {
        // 2. Handle Strings
        const s = String(dateStr).trim().toLowerCase();
        const months = ["jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"];
        
        for (let i = 0; i < months.length; i++) {
            if (s.includes(months[i])) {
                mIdx = i;
                break;
            }
        }

        // Improved year detection
        const yearMatch = s.match(/(20\d{2})/);
        if (yearMatch) {
            yIdx = parseInt(yearMatch[1]);
        } else {
            const shortYearMatch = s.match(/(\d{2})$/);
            if (shortYearMatch) {
                const yr = parseInt(shortYearMatch[1]);
                yIdx = 2000 + yr;
            } else {
                yIdx = fallbackYear;
            }
        }
    }

    if (mIdx !== -1 && yIdx !== -1) {
        return yIdx * 12 + mIdx;
    }
    return null;
}

function extractNumber(val: any): number | null {
    if (val === null || val === undefined || val === "") return null;
    if (typeof val === 'number') return val;
    
    const s = String(val).trim().toLowerCase();
    const zeros = ["-", "--", "( - )", "n.a.", "n/a", "tbd"];
    if (zeros.includes(s)) return 0.0;
    
    let cleaned = s.replace(/[$,£€%]/g, '');
    if (cleaned.startsWith('(') && cleaned.endsWith(')')) {
        cleaned = '-' + cleaned.slice(1, -1);
    }
    cleaned = cleaned.replace(/,/g, '');
    
    const num = parseFloat(cleaned);
    if (isNaN(num)) return null;
    return s.includes('%') ? num / 100.0 : num;
}

export function promoteByHierarchy(
    rows: any[][], 
    monthsCount: number, 
    mappings: Record<string, string>,
    startMonth: number = 0,
    startYear: number = 2024
): any {
    if (!rows || rows.length < 2) return {};

    const header = rows[0].map(c => String(c || "").trim());
    const dataRows = rows.slice(1);

    const roles: Record<string, number> = {};
    const hierarchyLevels: Array<{level: number, index: number}> = [];

    // 1. Identify Roles
    Object.entries(mappings).forEach(([idxStr, role]) => {
        const idx = parseInt(idxStr);
        if (role.startsWith("L")) {
            const level = parseInt(role.substring(1));
            if (!isNaN(level)) hierarchyLevels.push({ level, index: idx });
        } else {
            roles[role] = idx;
        }
    });

    hierarchyLevels.sort((a, b) => a.level - b.level);

    // 2. Format Detection (Wide vs Long) and Date Range bounds
    const hasMonthsInHeader = header.some(h => parseDateToAbsoluteMonth(h, startYear) !== null);
    const isLongFormat = "Value" in roles && ("Time" in roles || hasMonthsInHeader);

    const timeAbsValues: number[] = [];
    if (isLongFormat) {
        const timeIdx = roles["Time"];
        dataRows.forEach(row => {
            if (timeIdx !== undefined && timeIdx < row.length) {
                const absM = parseDateToAbsoluteMonth(row[timeIdx], startYear);
                if (absM !== null) timeAbsValues.push(absM);
            }
        });
    } else {
        header.forEach((h, i) => {
            if (hierarchyLevels.some(hl => hl.index === i) || i === roles["Scenario"]) return;
            const absM = parseDateToAbsoluteMonth(h, startYear);
            if (absM !== null) timeAbsValues.push(absM);
        });
    }

    let detectedDateRange: DateRange | null = null;
    if (timeAbsValues.length > 0) {
        const minAbs = Math.min(...timeAbsValues);
        const maxAbs = Math.max(...timeAbsValues);
        
        startYear = Math.floor(minAbs / 12);
        startMonth = minAbs % 12;
        const span = maxAbs - minAbs + 1;
        monthsCount = Math.max(span, monthsCount);
        
        const finalAbs = minAbs + monthsCount - 1;
        
        detectedDateRange = {
            startMonth,
            startYear,
            endMonth: finalAbs % 12,
            endYear: Math.floor(finalAbs / 12),
            monthsCount
        };
    }

    const startAbs = startYear * 12 + startMonth;

    // 3. Process Rows into a Hierarchy Map
    type KpiMapItem = { 
        id: string; 
        label: string; 
        levels: string[]; 
        scenarios: Record<string, number[]> 
    };
    const kpiMap: Record<string, KpiMapItem> = {};
    const getKpiKey = (levels: string[]) => levels.join(" > ");

    dataRows.forEach(row => {
        const currentLevels = hierarchyLevels
            .map(h => String(row[h.index] || "").trim())
            .filter(Boolean);
        
        if (currentLevels.length === 0) return;

        // Clean trailing blanks if any (already filter Boolean handled it but let's be safe)
        const kpiKey = getKpiKey(currentLevels);
        if (!kpiMap[kpiKey]) {
            const kpiId = getStableId(kpiKey);
            kpiMap[kpiKey] = {
                id: kpiId,
                label: currentLevels[currentLevels.length - 1],
                levels: currentLevels,
                scenarios: {}
            };
        }

        const scenarioName = roles["Scenario"] !== undefined ? String(row[roles["Scenario"]] || "Actual").trim() : "Actual";
        if (!kpiMap[kpiKey].scenarios[scenarioName]) {
            kpiMap[kpiKey].scenarios[scenarioName] = new Array(monthsCount).fill(0.0);
        }

        if (isLongFormat) {
            const val = extractNumber(row[roles["Value"]]) || 0.0;
            const timeIdx = roles["Time"];
            let mIdx = 0;
            if (timeIdx !== undefined) {
                const absM = parseDateToAbsoluteMonth(row[timeIdx], startYear);
                if (absM !== null) mIdx = absM - startAbs;
            }
            if (mIdx >= 0 && mIdx < monthsCount) {
                kpiMap[kpiKey].scenarios[scenarioName][mIdx] += val;
            }
        } else {
            // Wide format
            header.forEach((h, i) => {
                if (hierarchyLevels.some(hl => hl.index === i) || i === roles["Scenario"]) return;
                const absM = parseDateToAbsoluteMonth(h, startYear);
                if (absM !== null) {
                    const mIdx = absM - startAbs;
                    if (mIdx >= 0 && mIdx < monthsCount) {
                        const v = extractNumber(row[i]) || 0.0;
                        kpiMap[kpiKey].scenarios[scenarioName][mIdx] += v;
                    }
                }
            });
        }
    });

    // 4. Build Relationship Map and Registry
    const nodeRegistry: Record<string, any> = {};
    const sortedKpiKeys = Object.keys(kpiMap).sort((a, b) => a.split(" > ").length - b.split(" > ").length);

    // Ensure all levels exist in registry
    sortedKpiKeys.forEach(key => {
        const item = kpiMap[key];
        const levels = item.levels;
        for (let i = 1; i <= levels.length; i++) {
            const subPath = levels.slice(0, i);
            const subKey = getKpiKey(subPath);
            if (!nodeRegistry[subKey]) {
                const pPath = levels.slice(0, i - 1);
                const pKey = pPath.length > 0 ? getKpiKey(pPath) : null;
                const nodeId = getStableId(subKey);

                nodeRegistry[subKey] = {
                    id: nodeId,
                    label: subPath[subPath.length - 1],
                    unit: "$",
                    formula: i < levels.length ? 'SUM' : 'NONE',
                    children: [],
                    parentId: pKey ? nodeRegistry[pKey].id : undefined,
                    data: [], // populated per scenario
                    isExpanded: true,
                    simulationValue: 0,
                    simulationType: 'PERCENT'
                };
                
                if (pKey && nodeRegistry[pKey]) {
                    if (!nodeRegistry[pKey].children.includes(nodeId)) {
                        nodeRegistry[pKey].children.push(nodeId);
                    }
                }
            }
        }
    });

    // 5. Populate scenarios and aggregate
    const allScenarios = new Set<string>();
    Object.values(kpiMap).forEach(item => {
        Object.keys(item.scenarios).forEach(s => allScenarios.add(s));
    });
    if (allScenarios.size === 0) allScenarios.add("Actual");

    const scenariosOutput: Record<string, Record<string, KPIData>> = {};
    
    allScenarios.forEach(sName => {
        const sKpis: Record<string, KPIData> = {};
        
        // 5.1 Initialize leaf data
        Object.entries(nodeRegistry).forEach(([key, node]) => {
            const newNode = { ...node, children: [...node.children] };
            const vals = (kpiMap[key] && kpiMap[key].scenarios[sName]) || new Array(monthsCount).fill(0.0);
            
            newNode.data = vals.map((v: number, j: number) => ({
                month: (startYear * 12 + startMonth + j) % 12,
                year: Math.floor((startYear * 12 + startMonth + j) / 12),
                actual: v
            }));
            sKpis[newNode.id] = newNode;
        });

        // 5.2 Aggregation: Bottom-up
        const depthSortedKeys = Object.keys(nodeRegistry).sort((a, b) => b.split(" > ").length - a.split(" > ").length);
        depthSortedKeys.forEach(key => {
            const nodeId = nodeRegistry[key].id;
            const node = sKpis[nodeId];
            if (node.children && node.children.length > 0) {
                // If it's a SUM parent, recalculate from children
                if (node.formula === 'SUM') {
                    const totals = new Array(monthsCount).fill(0.0);
                    node.children.forEach(cid => {
                        const child = sKpis[cid];
                        if (child && child.data) {
                            child.data.forEach((d: any, idx: number) => {
                                if (idx < monthsCount) totals[idx] += d.actual || 0;
                            });
                        }
                    });

                    node.data = totals.map((v, j) => ({
                        month: ((startYear * 12 + startMonth + j) % 12).toString(),
                        year: Math.floor((startYear * 12 + startMonth + j) / 12),
                        actual: v
                    }));
                }
            }
        });

        scenariosOutput[sName] = sKpis;
    });

    // 6. Final Result
    const defaultScenario = scenariosOutput["Actual"] ? "Actual" : Object.keys(scenariosOutput)[0];
    const stats = {
        total_rows_processed: dataRows.length,
        kpis_created: Object.keys(kpiMap).length,
        scenarios_detected: Array.from(allScenarios)
    };

    const result = {
        type: 'multi-scenario',
        scenarios: scenariosOutput,
        kpis: scenariosOutput[defaultScenario],
        dateRange: detectedDateRange || { startMonth, startYear, monthsCount },
        stats
    };

    return result;
}

/**
 * Parses an Excel or CSV file in the browser and returns raw row data for all sheets.
 */
export async function parseFile(file: File): Promise<{ sheets: Record<string, any[][]>, firstSheetData: any[][] }> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onload = (e) => {
            const data = new Uint8Array(e.target?.result as ArrayBuffer);
            const workbook = XLSX.read(data, { type: 'array' });
            
            const sheets: Record<string, any[][]> = {};
            workbook.SheetNames.forEach(name => {
                const worksheet = workbook.Sheets[name];
                sheets[name] = XLSX.utils.sheet_to_json(worksheet, { header: 1 }) as any[][];
            });

            const firstSheetData = sheets[workbook.SheetNames[0]] || [];
            resolve({ sheets, firstSheetData });
        };
        reader.onerror = (err) => reject(err);
        reader.readAsArrayBuffer(file);
    });
}
